File: work/python3/binary-search/35_SearchInsertPosition.py
#
# @lc app=leetcode id=35 lang=python3
#
# [35] Search Insert Position
#

# @lc code=start
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        # find median index
        if len(nums)%2 == 0:
            cur_check=len(nums)//2-1
        else:
            cur_check=len(nums)//2
        
        if nums[0] > target:
            return 0
        elif nums[len(nums)-1] < target:
            return len(nums)
        else:
            pre_check=-1
            while True:
                if nums[cur_check] < target:
                    cur_check=cur_check+1
                    if pre_check == cur_check:
                        return cur_check
                    pre_check=cur_check-1
                elif nums[cur_check] > target:
                    cur_check=cur_check-1
                    if pre_check == cur_check:
                        return cur_check+1
                    pre_check=cur_check+1
                else:
                    return cur_check
        
# @lc code=end

File: work/typescript/array/35_SearchInsertPosition.ts
/*
 * @lc app=leetcode id=35 lang=typescript
 *
 * [35] Search Insert Position
 */

// @lc code=start
function searchInsert(nums: number[], target: number): number {
    let isPreLow:boolean|undefined;
    let i:number=Math.floor(nums.length/2);
    while(0<=i && i<nums.length){
        if(nums[i]<target){
            if(isPreLow==false){
                return i+1;
            }
            else{
                isPreLow=true;
                i++;
            };
        }
        else if(target<nums[i]){
            if(isPreLow===true){
                return i;
            }
            else{
                isPreLow=false;
                i--;
            };
        }
        else{
            return i;
        }
    };
    if(i==-1){
        return 0;
    }
    else{
        return i;
    }
};
// @lc code=end

